### ISSUE TITLE
Captcha detected

### ISSUE THREAD

**Issue by jairoxyz** [FlareSolverr/FlareSolverr#5]
Hi, this is a great project. I was actually looking into the same direction after CF changed their challenge in June but could not get it work with puppeteer headless chrome. Your solution works fine but after a while I get captcha detected errors. Any idea what might be causing this? In python based CF solvers trigger of captcha used to be caused by problematic TLSv1.0 ciphers that had to be removed from the urllib3 default cipher suite. Could this be similar with puppeteer?

Thanks,

Jx-

**Comment by NoahCardoza** [comment]
Which site is this happening on? I tend to get captcha's on my normal browser sometimes. I'm working on a [patch](https://github.com/NoahCardoza/FlareSolverr) that will include automatic captcha solving.

**Comment by jairoxyz** [comment]
Different sites, for example http://www.seehd.pl or https://yesmovies.ag/searching/joker.html. It maybe is related to the comments here https://github.com/ngosang/FlareSolverr/issues/4#issuecomment-652896084 about SSL and browser fingerprinting. Will try your fork if the session implementation (nice one:)) causes less captchas. Havent experience in normal browser on mentioned sites yet.

**Comment by jairoxyz** [comment]
PS: using the sessions I don't get captchas while in a session but then when expiring it and trying a new one the captcha is back :)

Also, I am running on Windows 10 and there seems to be a bug in fs.rmdirSync because the folder or some file in it is locked or something:
```
  errno: -4082,
  syscall: 'unlink',
  code: 'EBUSY',
  path: '\\\\?\\C:\\Users\\xyz\\AppData\\Local\\Temp\\puppeteer_firefox_profile_123\\cert9.db'
```
I tried fs-extra but the same. According to google this happens on Windows only. Gonna search for some other lib that works.

Cheers,

Jx-

**Comment by NoahCardoza** [comment]
I'm wondering if it has to do with it being a temp directory? Let me know what you find. It should run fine in the docker though.

**Comment by NoahCardoza** [comment]
@jairoxyz Just pushed captcha solving support. Haven't written up any documentation but if you take a look at the code it should make sense more or less. People will be able to write their own adapters to different services like 2cap, etc. And select which they'd like to use when building the container.

To start it off, I added a `harvester` adapter which works with my [CaptchaHarvester](https://github.com/NoahCardoza/CaptchaHarvester) project.

**Comment by jairoxyz** [comment]
@NoahCardoza great to add captcha solver support. Didn't know your harvester project yet. Looks excellent!

Re session folder clean up. Tried all sorts of approaches on Windows but those files are always locked. Will keep investigating.

On docker I get another error, maybe because the node version 10 used in the image? I think the recursive switch is pretty new feature of fs.rmdirsync.
```
syscall: 'rmdir',
  code: 'ENOTEMPTY',
  path: '/tmp/puppeteer_firefox_profile_000002'
```
Another issue I had was with 
`const { getCaptchaSolver } = require('./captcha').defualt`
had to change to
`const { getCaptchaSolver } = require('./captcha').getCaptchaSolver`

and then a minor cosmetic thing: you use getCaptchaSovler - not sure if that is a typo or intentional :) 

Cheers,

Jx-

**Comment by jairoxyz** [comment]
PS: recursive fs.rmdir needs node version 12.

I use this function and that works fine in the Docker image

```
const fs = require('fs');
const Path = require('path');

const deleteFolderRecursive = function(path) {
  if (fs.existsSync(path)) {
    fs.readdirSync(path).forEach((file, index) => {
      const curPath = Path.join(path, file);
      if (fs.lstatSync(curPath).isDirectory()) { // recurse
        deleteFolderRecursive(curPath);
      } else { // delete file
        fs.unlinkSync(curPath);
      }
    });
    fs.rmdirSync(path);
  }
};
```

**Comment by jairoxyz** [comment]
PPS: I am also getting captcha challenges with response 200, so maybe line 244 could be change like this?

`if (response.status() === 403 || html.includes("cf_captcha_kind")) {`

**Comment by NoahCardoza** [comment]
> @NoahCardoza great to add captcha solver support. Didn't know your harvester project yet. Looks excellent!
> 
> Re session folder clean up. Tried all sorts of approaches on Windows but those files are always locked. Will keep investigating.
> 
> On docker I get another error, maybe because the node version 10 used in the image? I think the recursive switch is pretty new feature of fs.rmdirsync.
> 
> ```
> syscall: 'rmdir',
>   code: 'ENOTEMPTY',
>   path: '/tmp/puppeteer_firefox_profile_000002'
> ```
> 
> Another issue I had was with
> `const { getCaptchaSolver } = require('./captcha').defualt`
> had to change to
> `const { getCaptchaSolver } = require('./captcha').getCaptchaSolver`
> 
> and then a minor cosmetic thing: you use getCaptchaSovler - not sure if that is a typo or intentional :)
> 
> Cheers,
> 
> Jx-

Good catch! Yeah, it was a typo. Not sure how the `.default` got there because that breaks it, but it was running for me I could have sworn right before I pushed...

Fixed it with https://github.com/NoahCardoza/FlareSolverr/commit/436837f94e84d976da7098c1f341734c805fba1c.

**Comment by NoahCardoza** [comment]
Ok, implemented your custom solution and added more checking for captchas!

**Comment by jairoxyz** [comment]
@NoahCardoza I added a try/catch block around the rmdir part, so that it doesnt crash the whole process on Windows, plus, I tried to have it wait a few secs before deleting the files and then it works. Maybe I do a pull request?

Also noted that when running on Docker, it hangs most of the time when calling the page a second time, either at "Navigating to page" or "Waiting for challenge". This doesn't happen when running with node. Maybe a matter of resources on the docker image VM?

Jx-

**Comment by NoahCardoza** [comment]
Sounds good! I added a few comments to the PR.

As for the Docker image, you could be right. I noticed it seems to hang sometimes, but never on "Navigating to page" as far as I can remember. It only happened to me when it hit the "Wait 5 seconds" pages. I added a reload after 8 or so seconds if I remember correctly and that seemed to fix it. I'll run some tests and see if I'm having the same issues.

Currently updating the README. I think I may move all the changes to my own repo since this one doesn't seem to be very active and I've made so many changes, the API is totally changed and it makes more sense to open the issues on the repo that actually has the code since forks don't get an issue's page sadly.

**Comment by jairoxyz** [comment]
@NoahCardoza yeah good idea to move all to your repo :)

Great code not with refactoring into modules and sessions commands ;)  With the current code, I did not get anymore hang-ups but one strange thing I noticed when running in docker. The session folder created has a different (random?) name than the one specified in the session param. Though the session seems to be reused correctly, its still strange and am not sure if this can cause problems.
